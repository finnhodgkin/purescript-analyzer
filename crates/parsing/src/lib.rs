pub mod grammar;
pub mod output;
pub mod parser;
