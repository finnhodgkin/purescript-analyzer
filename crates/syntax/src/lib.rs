pub mod ast;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[repr(u16)]
pub enum SyntaxKind {
    Whitespace = 0,
    LineComment,
    BlockComment,

    Module,
    ModuleHeader,
    ModuleKw,
    WhereKw,

    ExportList,
    ImportList,

    ImportDeclaration,
    ImportKw,
    AsKw,

    ModuleName,
    QualifiedName,
    QualifiedPrefix,
    Name,
    NameRef,
    Upper,
    Lower,
    Hole,
    Operator,

    Equal,
    Period,
    Period2,
    Colon,
    Colon2,
    LeftArrow,
    RightArrow,
    LeftThickArrow,
    RightThickArrow,
    LeftParenthesis,
    RightParenthesis,
    LeftBracket,
    RightBracket,
    LeftBrace,
    RightBrace,
    Tick,
    Comma,
    Pipe,

    ApplicationExpression,
    ConstructorExpression,
    LiteralExpression,
    OperatorChain,
    ParenthesizedExpression,
    TernaryExpression,
    TypedExpression,
    VariableExpression,

    LiteralChar,
    LiteralString,
    LiteralRawString,
    LiteralInteger,
    LiteralNumber,
    LiteralTrue,
    LiteralFalse,

    ConstructorType,
    VariableType,

    Pattern,

    ValueDeclaration,
    AnnotationDeclaration,

    DataDeclaration,
    DataKw,

    NewtypeDeclaration,
    NewtypeKw,
    ForallKw,

    TypeDeclaration,
    TypeKw,

    ClassDeclaration,
    ClassKw,

    InstanceDeclaration,
    InstanceKw,

    DeriveInstanceDeclaration,
    DeriveKw,

    ForeignDataDeclaration,
    ForeignValueDeclaration,
    ForeignKw,

    FixityDeclaration,
    InfixlKw,
    InfixrKw,
    InfixKw,

    Sentinel,
    Error,
    EndOfFile,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct PureScript {}

impl rowan::Language for PureScript {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        assert!(raw.0 <= SyntaxKind::EndOfFile as u16);
        unsafe { std::mem::transmute::<u16, SyntaxKind>(raw.0) }
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        rowan::SyntaxKind(kind as u16)
    }
}

impl From<SyntaxKind> for rowan::SyntaxKind {
    fn from(value: SyntaxKind) -> Self {
        Self(value as u16)
    }
}

pub type SyntaxNode = rowan::SyntaxNode<PureScript>;
pub type SyntaxNodeChildren = rowan::SyntaxNodeChildren<PureScript>;
pub type SyntaxToken = rowan::SyntaxToken<PureScript>;
pub type SyntaxElement = rowan::SyntaxElement<PureScript>;

impl SyntaxKind {
    pub fn is_whitespace_or_comment(&self) -> bool {
        matches!(self, Self::Whitespace | Self::LineComment | Self::BlockComment)
    }
}

#[cfg(test)]
mod tests {
    use rowan::{ast::AstNode, NodeOrToken};

    use crate::{ast, SyntaxElement, SyntaxKind, SyntaxNode};

    fn print(indent: usize, element: SyntaxElement) {
        let kind: SyntaxKind = element.kind();
        print!("{:indent$}", "", indent = indent);
        match element {
            NodeOrToken::Node(node) => {
                println!("- {:?}", kind);
                for child in node.children_with_tokens() {
                    print(indent + 2, child);
                }
            }
            NodeOrToken::Token(token) => println!("- {:?} {:?}", token.text(), kind),
        }
    }

    #[test]
    fn syntax_playground() {
        let mut builder = rowan::GreenNodeBuilder::default();

        builder.start_node(SyntaxKind::Module.into());
        builder.start_node(SyntaxKind::ModuleHeader.into());
        builder.token(SyntaxKind::ModuleKw.into(), "module");
        builder.token(SyntaxKind::Whitespace.into(), " ");
        builder.start_node(SyntaxKind::ModuleName.into());
        builder.token(SyntaxKind::Upper.into(), "PureScript");
        builder.token(SyntaxKind::Period.into(), ".");
        builder.token(SyntaxKind::Upper.into(), "Main");
        builder.finish_node();
        builder.token(SyntaxKind::Whitespace.into(), " ");
        builder.start_node(SyntaxKind::ExportList.into());
        builder.token(SyntaxKind::LeftParenthesis.into(), "(");
        builder.token(SyntaxKind::RightParenthesis.into(), ")");
        builder.finish_node();
        builder.token(SyntaxKind::Whitespace.into(), " ");
        builder.token(SyntaxKind::WhereKw.into(), "where");
        builder.finish_node();
        builder.finish_node();

        let purescript_module = SyntaxNode::new_root(builder.finish());

        println!("{}", purescript_module);
        print(2, purescript_module.clone().into());

        let module_name = purescript_module
            .children()
            .nth(0)
            .unwrap()
            .children()
            .nth(0)
            .and_then(ast::ModuleName::cast)
            .unwrap();

        let rust_module = SyntaxNode::new_root(
            module_name
                .segments()
                .nth(0)
                .unwrap()
                .replace_with(rowan::GreenToken::new(SyntaxKind::Upper.into(), "Rust")),
        );

        println!("{}", rust_module);
        print(2, rust_module.into());
    }
}
