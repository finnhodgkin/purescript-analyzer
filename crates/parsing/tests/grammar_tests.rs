use std::fmt::Write;

use parsing::{
    grammar::{expression, qualified_name},
    lexer::lex,
    parser::{Event, Parser},
};
use syntax::SyntaxKind;

fn expect_parse<F, T>(name: &str, source: &str, rule: F)
where
    F: Fn(&mut Parser) -> T,
{
    let lexed = lex(source);
    let input = lexed.as_input();
    let mut parser = Parser::new(input);
    let _ = rule(&mut parser);

    let mut result = String::new();
    writeln!(result, "Input: {source}").unwrap();

    let mut indentation = 0;
    for actual in parser.as_output() {
        if let Event::Start { kind: SyntaxKind::Sentinel } = actual {
            continue;
        }
        match actual {
            Event::Start { .. } => {
                writeln!(result, "{:indentation$}{:?}", "", actual, indentation = indentation)
                    .unwrap();
                indentation += 2
            }
            Event::Finish => {
                indentation -= 2;
                writeln!(result, "{:indentation$}{:?}", "", actual, indentation = indentation)
                    .unwrap();
            }
            _ => {
                writeln!(result, "{:indentation$}{:?}", "", actual, indentation = indentation)
                    .unwrap();
            }
        }
    }

    insta::assert_snapshot!(name, result);
}

#[test]
fn test_qualified_constructor() {
    expect_parse("qualified-constructor", "Hello.World", qualified_name);
}

#[test]
fn test_qualified_variable() {
    expect_parse("qualified-variable", "Hello.World.hello", qualified_name);
}

#[test]
fn test_qualified_variable_as() {
    expect_parse("qualified-variable-as", "Hello.World.as", qualified_name);
}

#[test]
fn test_qualified_name_error() {
    expect_parse("qualified-error", "Hello.World.(import", qualified_name);
}

#[test]
fn test_qualified_name_plural() {
    expect_parse("qualified-plural", "(hello) world", |parser| {
        qualified_name(parser);
        qualified_name(parser);
    })
}

#[test]
fn test_expression() {
    expect_parse("literal-int", "1", expression);
    expect_parse("operator-chain", "1 + 2 * 3", expression);
    expect_parse("application-expression", "1 2", expression);
    expect_parse("typed-expression", "1 :: Int", expression);
    expect_parse("parenthesized-expression", "(1)", expression);
    expect_parse("parenthesized-application", "(1 2)", expression);
    expect_parse("parenthesized-operator-application", "(1 2) + (3 4)", expression);
}
