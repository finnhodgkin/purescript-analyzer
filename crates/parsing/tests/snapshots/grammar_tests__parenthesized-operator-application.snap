---
source: crates/parsing/tests/grammar_tests.rs
expression: result
---
Input: (1 2) + (3 4)
Start { kind: OperatorChain }
  Start { kind: ParenthesizedExpression }
    Token { kind: LeftParenthesis }
    Start { kind: ApplicationExpression }
      Start { kind: LiteralExpression }
        Token { kind: LiteralInteger }
      Finish
      Start { kind: LiteralExpression }
        Token { kind: LiteralInteger }
      Finish
    Finish
    Token { kind: RightParenthesis }
  Finish
  Token { kind: Operator }
  Start { kind: ParenthesizedExpression }
    Token { kind: LeftParenthesis }
    Start { kind: ApplicationExpression }
      Start { kind: LiteralExpression }
        Token { kind: LiteralInteger }
      Finish
      Start { kind: LiteralExpression }
        Token { kind: LiteralInteger }
      Finish
    Finish
    Token { kind: RightParenthesis }
  Finish
Finish

